### PR TITLE
Updated generator, build system, readme, sample data and images

### PR DESCRIPTION
**Pull request by bernardboey**
This pull request transfers over the updates worked over the summer that includes the following changes:

**1) Build system uses Makefile and configure**
This ensures support for various Linux distributions. It is also a more widespread build system.

**2) Cartogram generator reads GeoJSON and CSV**
GeoJSON is a more up-to-date format for representing geographic data. CSV is a more prevalent format for representing statistics. Support for .gen and .dat still remains. However, mention of this is excluded in the README for sake of simplicity.

**3) US sample data changed to GeoJSON and CSV**
Note that India and China is removed.

**4) New flag "-p" for processing GeoJSON files**
The cartogram generator relies on a "cartogram_id" property present in the GeoJSON file to match the id with the region name. To settle this, a user can run `cartogram -p [map_file_name]` in order for the program to insert the "cartogram_id" properties into the GeoJSON file. A CSV file is also produced which contains the region names and matching IDs. This replaces the need for an external script to process the file. Do note that this assumes that the GeoJSON file already contains a region name property.

**5) Missing data can be represented in the cartogram**
By typing "NA" instead of a number in the CSV file, the generator greys out the region with missing data and ensures that its relative area remains the same. Do note that the shape may still be distorted to accommodate the neighbouring regions.

**6) Extremely small values are rounded up**
The program calculates a threshold for each region based on its perimeter. If the relative target area is below this threshold, the target area is altered and set to the threshold area. This ensures that the regions do not get too small, which may be visually undesirable. In addition, this speeds up the calculation of the cartogram as extremely small values slow down the program.

**7) README reflects the above changes and sample images are updated**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,18 +9,21 @@
...
 
-<hr>
-
-[![release](http://github-release-version.herokuapp.com/github/Flow-Based-Cartograms/go_cart/release.svg?style=flat)](https://github.com/Flow-Based-Cartograms/go_cart/releases/latest)
-[![Maintenance](https://img.shields.io/badge/maintained-yes-green.svg)](https://GitHub.com/Naereen/StrapDown.js/graphs/commit-activity)
-[![HitCount](http://hits.dwyl.io/Flow-Based-Cartograms/go_cart.svg)](http://hits.dwyl.io/Flow-Based-Cartograms/go_cart)
-[![Tweet](https://img.shields.io/twitter/url/http/shields.io.svg?style=social)](https://twitter.com/intent/tweet?url=https%3A%2F%2Fgithub.com%2FFlow-Based-Cartograms%2Fgo_cart&text=Create%20your%20own%20cartograms%20today%20-%20and%20see%20what%20your%20world%20really%20looks%20like%21&hashtags=data%2Cmap%2Ccartogram%2Ctech)
-
-
+## Introduction
```
Comment (pratyushmore) [inline README.md:29]: Why did you remove the GitHub tags?

File: README.md [hunk 2]
```diff
@@ -41,34 +45,25 @@ Gastner, M., Seguy, V., & More, P. (2018). Fast flow-based algorithm for creatin
...
-    ...  
-    0.334699 -0.204771  
-    END                       Each polygon terminates with END.  
-    END                       One more END signals the end of the file.
-    ```
-
-2. a `.dat` file containing the data (such as population) for each region, according to which these will be scaled. For the 2016 US presidential election data we provide `usa_electors.dat` (also in the `sample_data` folder) which provides the number of electors for each state.
-
-<!-- We can add some explanation about the files here if needed -->
+The cartogram generator expects two input file. You will need to obtain the '.json' file. The '.csv' file will be generated by the cartogram generator from the '.json' file. If you do not have a '.json' file, please follow the steps listed here: https://github.com/Flow-Based-Cartograms/go_cart/blob/master/process_map_file.md
```
Comment (pratyushmore) [inline README.md]: This part of the Readme is not clear. You mention a JSON and CSV file before introducing them. I would rework this.

File: README.md [hunk 3]
```diff
@@ -84,41 +79,55 @@ No additional dependencies. Your default `apt-get` package manager should work f
 
 1. Open a terminal, clone the repository, and navigate to its root directory.
 
-2. Run the provided automatic build script (you may need to grant sudo priveleges):
+2. Run the provided automatic build script. This will install the required dependencies and run configure and make for you:
+```
+sudo bash autobuild.sh
```
Comment (pratyushmore) [inline README.md]: Why change this? Some shells might not support `bash`.

File: README.md [hunk 4]
```diff
@@ -84,41 +79,55 @@ No additional dependencies. Your default `apt-get` package manager should work f
...
+```
+
+3. The `cartogram` executable can be found under `cartogram_generator/`. To install it, run the following command:
```
Comment (pratyushmore) [inline README.md:95]: Why not copy it to the root directory, as before? Is there a need to change this?

File: README.md [hunk 5]
```diff
@@ -84,41 +79,55 @@ No additional dependencies. Your default `apt-get` package manager should work f
...
 ```
-chmod a+x autobuild.sh && ./autobuild.sh
+sudo make install
```
Comment (pratyushmore) [inline README.md:97]: What is the installation for? Are there dynamically linked libraries?

File: README.md [hunk 6]
```diff
@@ -84,41 +79,55 @@ No additional dependencies. Your default `apt-get` package manager should work f
...
+-e: generate cartogram figure in EPS format
+-i: calculate inverse transform
+```
 
 **Note:** If you run into issues, look at the [**Troubleshooting**](#troubleshooting) section below.
 
 ## Running the cartogram generator
 
-1. Navigate to the directory with the `cartogram` executable. By default, this should be the root directory of the project.
+First, you need to process the '.json' file. If you do not have a '.json' file, please follow the tutorial here: https://github.com/Flow-Based-Cartograms/go_cart/blob/master/process_map_file.md
```
Comment (pratyushmore) [inline README.md]: The instructions to obtain files should be in the `Input Data Format` section.

File: README.md [hunk 7]
```diff
@@ -130,53 +139,58 @@ evince <filename>.eps
...
-For our 2016 US election example, `map.eps` should look like:
+For our example, `map.eps` should look as follows:
 
 <p align="center">
-<img src ="sample_images/US_map.png" width="65%">
+<img src ="sample_images/usa_map.png" width="65%">
 </p>
 
-`cartogram.eps` should look like:
+And here is how `cartogram.eps` should look like:
```
Comment (pratyushmore) [inline README.md]: nit: delete `here is how`

File: README.md [hunk 8]
```diff
@@ -84,41 +79,55 @@ No additional dependencies. Your default `apt-get` package manager should work f
...
+
+3. The `cartogram` executable can be found under `cartogram_generator/`. To install it, run the following command:
 ```
-chmod a+x autobuild.sh && ./autobuild.sh
+sudo make install
 ```
 
-3. You should have an executable by the name of `cartogram` in your root directory.
+4. Once you have installed it, you can run the cartogram generator at the command line by running `cartogram`:
+```
```
Comment (pratyushmore) [inline README.md:101]: suggestion: "... `cartogram`. You should see the following output:"

File: README.md [hunk 9]
```diff
@@ -130,53 +139,58 @@ evince <filename>.eps
...
 
-## Troubleshooting
-### Error: Don't understand 'm' flag!
+### Options
 
-If you see the following line in your output, you may want to take a look at <a href="https://github.com/dmlc/xgboost/issues/1945" target="\_blank">this solution</a>. Once you have implemented the solution from the link, follow the standard build instructions above.
-```
-FATAL:/opt/local/bin/../libexec/as/x86_64/as: I don't understand 'm' flag!
-```
+**-e**<br/>
```
Comment (pratyushmore) [inline README.md:164]: What is the alternative?

File: README.md [hunk 10]
```diff
@@ -130,53 +139,58 @@ evince <filename>.eps
...
+
+## Troubleshooting
 In case you run into problems while building the generator using the automated script, you could try manually carrying out some of the steps.
 
 #### macOS
 1. Install fftw and gcc.
 ```
-brew install fftw && brew link fftw
-brew install gcc
+$ brew install fftw
```
Comment (pratyushmore) [inline README.md:189]: Why change this?

File: autobuild.sh [hunk 11]
```diff
@@ -1,27 +1,11 @@
 # Checking for Operating System
 if [ "$(uname)" == "Darwin" ]; then
     # Install dependencies and Generate Makefile on Mac
-    chmod a+x scripts/prepare_env_mac.sh && scripts/prepare_env_mac.sh
+    bash scripts/build_mac.sh
```
Comment (pratyushmore) [inline autobuild.sh:4]: Why change this?

File: autobuild.sh [hunk 12]
```diff
@@ -1,27 +1,11 @@
...
 elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]; then
     # Install dependencies and Generate Makefile on Linux
-    chmod a+x scripts/prepare_env_linux.sh && scripts/prepare_env_linux.sh
-#elif [ "$(expr substr $(uname -s) 1 10)" == "MINGW32_NT" ]; then
-    # Do something under 32 bits Windows NT platform
-#elif [ "$(expr substr $(uname -s) 1 10)" == "MINGW64_NT" ]; then
-    # Do something under 64 bits Windows NT platform
+    bash scripts/build_linux.sh
```
Comment (pratyushmore) [inline autobuild.sh:7]: ditto

**Review by pratyushmore — COMMENTED** [review]
Hi @bbkc22113 thank you so much for this contribution!

I've added a few comments - mostly on the README, or clarifications.

Also, I see a lot of files under the `.deps`. I am guessing this is to do with the new build system you introduced. Could you please elaborate on the new build system, and what purpose these files serve?

Once all this is responded to, I will also test the code, and if all works, merge it in.

Again, we really appreciate this great contribution!

**Comment by bernardboey** [comment]
Hi Pratyush, thanks for looking through it!

I have responded to most of the comments and made additional commits to rectify the issues you brought up. However, there are some questions which involve the new build system. I am not the best person to answer as Ian Duncan was the one who implemented it. I have forwarded your questions to him and I will post them here once he replies!

In the meantime, let me know whether the rest (excluding build system) are okay!

File: README.md (hunk 1 above)
Comment (bernardboey) [reply README.md:29]: This was an accident! The private repository did not have them and when transferring over I forgot to include them. This has been fixed in a new commit.

File: README.md (hunk 3 above)
Comment (bernardboey) [reply README.md]: Changed to `sudo ./autobuild.sh`

File: README.md (hunk 7 above)
Comment (bernardboey) [reply README.md]: Fixed

File: README.md (hunk 8 above)
Comment (bernardboey) [reply README.md:101]: Added

File: README.md (hunk 9 above)
Comment (bernardboey) [reply README.md:164]: Without the -e flag, there will be no image output. There will only be GeoJSON output which contains the geographical coordinates for the cartogram. Would you like some mention of this in the README?

File: README.md (hunk 6 above)
Comment (bernardboey) [reply README.md]: Removed

Comment (bernardboey) [reply README.md]: (Because it is a duplicate of info in the `Input Data Format` section)

File: README.md (hunk 2 above)
Comment (bernardboey) [reply README.md]: Changed to this:

The cartogram generator expects two input file.

1. a '.json' file containing the cartesian coordinates for the map regions under consideration. For example, for the 2016 US presidential election data we provide `usa_contiguous_arcgis.json` (in the `sample_data` folder) which includes the coordinates for the boundaries of the different states of the contiguous United States. You will need to obtain the '.json' file and it needs to adhere to the GeoJSON specifications. It also needs to be processed first by using the command `cartogram -p [json_file_name]` (Instructions found below).  If you do not have a '.json' file, please follow the steps listed here: https://github.com/Flow-Based-Cartograms/go_cart/blob/master/process_map_file.md

2. a '.csv' file containing the data (such as population) for each region, according to which these will be scaled. For the 2016 US presidential election data we provide `usa_contiguous_electors.csv` (in the `sample_data` folder) which provides the number of electors for each state. The csv file will be generated by the cartogram generator when you process the '.json' file. You will then need to add in the data to the csv file.

File: README.md (hunk 4 above)
Comment (jansky) [reply README.md:95]: By default the Makefile generated by `configure` places the built executable in the source code directory. This has been okay for me, but we can modify `autobuild.sh` to copy the executable like before.

File: README.md (hunk 5 above)
Comment (jansky) [reply README.md:97]: Installation only puts the `cartogram` executable in `/usr/bin/` or `/usr/local/bin`, so it's on the system path. I think it makes it more convenient to run the cartogram generator, but it's not a necessary step.

File: README.md (hunk 10 above)
Comment (jansky) [reply README.md:189]: I'm not sure. This can probably be reverted to the original.

File: autobuild.sh (hunk 11 above)
Comment (jansky) [reply autobuild.sh:4]: The scripts used to only install software and set some environment variables, hence the name `prepare_env`. With the new build system, I've found it more convenient to run `configure` and `make` within each OS-specific shell script. This is mainly because `configure` has to be invoked differently on Mac OS to tell it to use the appropriate version of GCC.

**Review by jansky — COMMENTED** [review]
Hi Pratyush,

I have answered some of your questions about changes related to the new build system. Bernard also let me know that you had an additional question:

> Also, I see a lot of files under the .deps. I am guessing this is to do with the new build system you introduced. Could you please elaborate on the new build system, and what purpose these files serve?

The new build system is designed to allow the cartogram generator to be more easily built on a wide variety of systems, and to be easier to configure. It uses GNU Autotools, which has been in widespread use for open-source C and C++ projects for many years. I retained the `autobuild.sh` script to automate installing dependencies on Mac OS and Ubuntu, and then running `configure` and `make`. The main benefits of Autotools are the following:

- Autotools looks for project dependencies (like FFTW or OpenMP), and generates Makefiles with system-specific CFLAGS and LDFLAGS automatically. If a prerequisite is not found, an error message describing the problem is printed in plain English. This was the original motivator for switching build systems, because we had some issues getting GCC to find the location of the FFTW header files on Mac OS.
- Autotools works on all UNIX-like operating systems, and Windows with a UNIX compatibility layer like Cygwin or the new Windows Subsystem for Linux
- Autotools supports out-of-tree builds, allowing you to build the project multiple times, separately. For example, you might have one build for your Linux system, another build for Windows using a cross compiler (I've done this to get a working Windows build), and a statically-built Linux binary intended for deployment on a service like Amazon AWS Lambda
- Autotools makes it easy to specify different build options, like building a statically-linked executable instead of a dynamically-linked one.
- Autotools-generated makefiles come with some nice targets for free, like `distcheck`, which produces a tarball of the repository contents after making sure it compiles without issues.

The new files for the build system are:

- `configure.ac` - This tells Autotools how to generate the `configure` script, which tests for project dependencies and sets compiling and linking flags
- `Makefile.am` - This tells Autotools that our source code is in the `cartogram_generator/` directory
- `cartogram_generator/Makefile.am` - This tells Autotools the source files for our project, and that we want to use them to build an executable

There are some additional files in the repository generated by Autotools from the above source files:
- `config/` - Some helper shell scripts used by `configure`.
- `Makefile.in` - Autotools produces this from our `Makefile.am`. The `Makefile.in` is used by `configure` to generate the actual `Makefile` invoked by the user to compile the project.

There is some debate about whether these generated files should be tracked in version control. If you want to untrack them, that's okay with me.

File: README.md (hunk 3 above)
Comment (pratyushmore) [reply README.md]: This will not work either. The file needs to be given executable perms. Please change back to the original command.

File: README.md (hunk 9 above)
Comment (pratyushmore) [reply README.md:164]: Yes, please. Each option should be described well.

File: autobuild.sh (hunk 11 above)
Comment (pratyushmore) [reply autobuild.sh:4]: I suppose my question was more around using the bash command vs the chmod and then executing.

**Comment by pratyushmore** [comment]
Hi @jansky and @bbkc22113 thanks for the updates, and apologies for the delayed response. 

I've responded to some of your questions. 

@jansky: As for the build system, if certain files are generated based on configs, IMO, these should not be tracked. They should be regenerated as part of the install script. Please untrack them, and add these files to the .gitignore file. 

Thank you!

File: README.md (hunk 1 above)
Comment (pratyushmore) [reply README.md:29]: Okay! I would suggest, that after this PR has been merged in, you fork this project, and make changes to the fork. That way we will not see these drifts, and unwanted changes.

**Comment by pratyushmore** [comment]
Closing this PR since a new one has been opened.

I've commented on the new PR. Moving forward, please update that PR with responses/changes to the comments. Do not open a new PR.
